internal/controllers/gateway/route_utils.go

```diff
@@ -84,7 +86,62 @@ func getSupportedGatewayForRoute(ctx context.Context, mgrc client.Client, obj cl
 		// if the GatewayClass matches this controller we're all set and this controller
 		// should reconcile this object.
 		if gatewayClass.Spec.ControllerName == ControllerName {
-			gateways = append(gateways, &gateway)
+			allowedNamespaces := make(map[string]interface{})
+			// set true if we find any AllowedRoutes. there may be none, in which case any namespace is permitted
+			filtered := false
+			for _, listener := range gateway.Spec.Listeners {
+				// Only match if the listener type matches the protocol type
+				switch obj.(type) {
+				case *gatewayv1alpha2.HTTPRoute:
+					if !(listener.Protocol == gatewayv1alpha2.HTTPProtocolType || listener.Protocol == gatewayv1alpha2.HTTPSProtocolType) {
+						continue
+					}
+				case *gatewayv1alpha2.TCPRoute:
+					if listener.Protocol != gatewayv1alpha2.TCPProtocolType {
+						continue
+					}
+				case *gatewayv1alpha2.UDPRoute:
+					if listener.Protocol != gatewayv1alpha2.UDPProtocolType {
+						continue
+					}
+				case *gatewayv1alpha2.TLSRoute:
+					if listener.Protocol != gatewayv1alpha2.TLSProtocolType {
+						continue
+					}
+				default:
+					continue
+				}
+				if listener.AllowedRoutes != nil {
+					filtered = true
+					if *listener.AllowedRoutes.Namespaces.From == gatewayv1alpha2.NamespacesFromAll {
+						// we allow "all" by just stuffing the namespace we want to find into the map
+						allowedNamespaces[obj.GetNamespace()] = nil
+					} else if *listener.AllowedRoutes.Namespaces.From == gatewayv1alpha2.NamespacesFromSame {
+						allowedNamespaces[gateway.ObjectMeta.Namespace] = nil
+					} else if *listener.AllowedRoutes.Namespaces.From == gatewayv1alpha2.NamespacesFromSelector {
+						namespaces := &corev1.NamespaceList{}
+						selector, err := metav1.LabelSelectorAsSelector(listener.AllowedRoutes.Namespaces.Selector)
+						if err != nil {
+							return nil, fmt.Errorf("failed to convert LabelSelector to Selector for gateway %s",
+								gateway.ObjectMeta.Name)
+						}
+						err = mgrc.List(ctx, namespaces,
+							&client.ListOptions{LabelSelector: selector})
+						if err != nil {
+							return nil, fmt.Errorf("could not fetch allowed namespaces for gateway %s",
+								gateway.ObjectMeta.Name)
+						}
+						for _, allowed := range namespaces.Items {
+							allowedNamespaces[allowed.ObjectMeta.Name] = nil
+						}
+					}
+				}
+			}
+
+			_, allowedNamespace := allowedNamespaces[obj.GetNamespace()]
+			if !filtered || allowedNamespace {
+				gateways = append(gateways, &gateway)
+			}
```

Comment:
in the above sections we perform the following steps for any route object we are processing:

1. check the type and verify whether that type is supported
2. if the type is supported for the listener, verify the namespace if restrictions are in place

I think we should refactor this into a few discreet functions as the current code is slightly deep.